runtime/compiler/optimizer/JProfilingValue.cpp

```diff
@@ -222,75 +224,103 @@ TR_JProfilingValue::removeRedundantProfilingValueCalls()
-       (node->getSymbol()->getMethodSymbol()->isVirtual() ||
-        node->getSymbol()->getMethodSymbol()->isInterface()))
-      addProfiling(node->getFirstChild(), tt);
+   TR::TreeTop *preceedingTT = NULL;
+   TR::Node *profiledNode = NULL;
+   TR::Node *constNode = NULL;
+   if (node->getOpCode().isCallIndirect() && !node->getByteCodeInfo().doNotProfile() &&
+         (node->getSymbol()->getMethodSymbol()->isVirtual() || 
+            node->getSymbol()->getMethodSymbol()->isInterface()))
+      {
+      /**
+       * RecompilationModifier adds the profiling placeholder call for Virtual calls. 
+       * Quickly check the next treetop, if it is profiling same virtual call target. In that case we do not need to add
+       * Profiling Call as it is already there.
+       */
+      profiledNode = node->getFirstChild();
+      TR::Node *nextTTNode = cursor->getNextTreeTop() ? cursor->getNextTreeTop()->getNode() : NULL;
+      if (!(alreadyProfiledValues->isSet(profiledNode->getGlobalIndex()) || (nextTTNode && nextTTNode->isProfilingCode() &&
+                                                            nextTTNode->getOpCodeValue() == TR::treetop &&
+                                                            nextTTNode->getFirstChild()->getOpCode().isCall() &&
+                                                            (nextTTNode->getFirstChild()->getSymbolReference()->getReferenceNumber() == TR_jProfile32BitValue ||
+                                                               nextTTNode->getFirstChild()->getSymbolReference()->getReferenceNumber() == TR_jProfile64BitValue) &&
+                                                            nextTTNode->getFirstChild()->getNumChildren() == 3 && 
+                                                            nextTTNode->getFirstChild()->getFirstChild() == profiledNode)))
+         {
+         preceedingTT = cursor;
+         constNode = TR::Node::iconst(node, 2);
+         performTransformation(comp(), "%s Adding JProfiling PlaceHolder call to profile, virtual call node n%dn profiling n%dn\n",
+            optDetailString(), node->getGlobalIndex(), profiledNode);
+         }
+      }
    else if (!node->getByteCodeInfo().doNotProfile() && 
-       (node->getOpCodeValue() == TR::instanceof ||
-        node->getOpCodeValue() == TR::checkcast))
-      addVFTProfiling(node->getFirstChild(), tt->getPrevTreeTop(), true);
-
+               (node->getOpCodeValue() == TR::instanceof || node->getOpCodeValue() == TR::checkcast || node->getOpCodeValue() == TR::checkcastAndNULLCHK))
+      {
+      if (!alreadyProfiledValues->isSet(node->getFirstChild()->getGlobalIndex()))
+         {
+         preceedingTT = cursor->getPrevTreeTop();
+         profiledNode = TR::Node::createWithSymRef(node, TR::aloadi, 1, node->getFirstChild(),
+                                                   getSymRefTab()->findOrCreateVftSymbolRef());
+         constNode = TR::Node::iconst(node, NEED_NULLTEST);
```

Comment:
@fjeremic  Previously for profiling VFT pointers, it was analyzing nodes while walking tree-tops to identify candidate and add profiling trees to profile pointer. In this approach, it would know when to add NULL Test (For example, profiling object class of instanceOf/checkCast node's first child). This was designed for non profiling bodies. 
For profiling compilations, approach was to walk tree tops to look for profiling placeholder calls added by other optimizations and lower them by adding profiling trees. But we also need to profile class pointers (Virtual Calll/instanceOf/checkCast) and in case of instanceOf/checkCast we also need to add NULL TEST. 
Now to the usage of `NEED_NULLTEST` , previously, we had total two treetop walks. First treetop walk is for profiling compilations which removes redundant profiling candidates in the same walk, I added code to scan VFT profiling candidates and add a placeholder calls. Now for placeholder calls, we have unused third child `iconst` which I am utilizing here to tell the next tree top walk that it needs NULL test. So `NEED_NULLTEST` is used to just increase readability.

Reply:
This seems a bit non-intuitive to me - none of the jProfile<n>BitValues will survive because the implementation is always in trees. Having varying numbers of children with iconsts to distinguish them is not making it clear what we need to do. Why not just add a second helper jProfile<n>BitValueWithNULLCHK or something and use that for the ones that need the special null handling...

Reply:
Main purpose of different number of children here is to identify placeholder calls which are as identified expanded to actual profiling code. Call Symbol for place holder calls and actual helper call to add new profiling value to table at runtime is same and we add actual helper calls in this opt pass while walking tree tops to lower place holder calls. So as dummy placeholder call already has third child, I am using it to serve my purpose as well.
I agree with non-intuitiveness with this. One thing I can do to increase intuitiveness is to create Enum for the constants and wherever we add placeholder calls, make changes to use Enum instead of non-intuitive constant values.

Reply:
Other option is to have a separate helper which is not implemented and doesn't survive to codegen to represent the placeholders which are then lowered into the real calls so you don't even need the enum

Reply:
Going to add dummy two child helpers one for NULLCHK one and other for normal one.

Reply:
Added new Symbols to represent placeholder calls in https://github.com/eclipse/omr/pull/4172  and also made corresponding changes as part of this PR.